File: app/javascript/channels/room_channel.js
import consumer from "./consumer"

$(function() {
  var createRoomChannel;

  jQuery(document).on('turbolinks:load', function() {
    var messages;
    messages = $('#messages');
    if (messages.length > 0) {
      createRoomChannel(messages.data('room-id'));
    }
    return $(document).on('keypress', '#message_body', function(event) {
      var message;
      message = event.target.value;
      if (event.keyCode === 13 && message !== '') {
        App.room.speak(message);
        event.target.value = "";
        return event.preventDefault();
      }
    });
  });

  createRoomChannel = function(roomId) {
    return App.room = consumer.subscriptions.create({
      channel: "RoomChannel",
      roomId: roomId
    }, {
      connected: function() {
        return console.log('Connected to RoomChannel');
      },
      disconnected: function() {
        return console.log('Disconnected from RoomChannel');
      },
      received: function(data) {
        console.log('Received message: ' + data['message']);
        return $('#messages').append(data['message']);
      },
      speak: function(message) {
        return this.perform('speak', {
          message: message
        });
      }
    });
  };

}).call(this);

File: app/channels/room_channel.rb
class RoomChannel < ApplicationCable::Channel
  def subscribed
    logger.info "Subscribed to RoomChannel, roomId: #{params[:roomId]}"

    @room = Room.find(params[:roomId])

    stream_from "room_channel_#{@room.id}"

    speak('message' => '* * * joined the room * * *')
  end

  def unsubscribed
    logger.info "Unsubscribed to RoomChannel"

    speak('message' => '* * * left the room * * *')
  end

  def speak(data)
    logger.info "RoomChannel, speak: #{data.inspect}"

     MessageService.new(
        body: data['message'], user: current_user, room: @room
     ).perform
  end
end
